### PR TITLE
[1LP][RFR] Entrypoints addition

### PR DESCRIPTION
**Pull request by psav**
This is a large change to the way providers are registered now

* Removed the add_provider_type method from the BaseProvider class
* Removed the add_base_type method from the BaseProvider class
* Added base_types as a function to the cfme.common.provider module
* Added provider_types as a function to the cfme.common.provider module
* Added entrypoints to the setup.py file
* Added load_setuptools_entrypoints to the utils.providers module

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit abae94385b0afe9c8bb486f11c4550a7e51c5935:

`utils/providers.py`:
- :grey_exclamation: Line 49:25: **S001** *found %r formatter*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit d20198c72725d5b17746f0bc249e6215b96a6ec7:

`cfme/common/provider.py`:
- :red_circle: Line 2:1: **F401** *'pkgutil' imported but unused*
- :red_circle: Line 3:1: **F401** *'importlib' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit bb03c62955f520b97b5676b67c6eb39017b322b0:

`cfme/common/provider.py`:
- :red_circle: Line 2:1: **F401** *'pkgutil' imported but unused*
- :red_circle: Line 3:1: **F401** *'importlib' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/common/provider.py [hunk 1]
```diff
@@ -51,13 +60,11 @@ class BaseProvider(Taggable, Updateable, SummaryMixin, Navigatable):
 
     @classmethod
     def add_base_type(cls, nclass):
-        cls.base_types[nclass.category] = nclass
-        return nclass
+        pass
```
Comment (RonnyPfannschmidt) [inline cfme/common/provider.py]: may be useful to add a deprecation warning here and one below

File: utils/providers.py [hunk 2]
```diff
@@ -380,9 +396,9 @@ def setup_a_provider(prov_class="infra", prov_type=None, validate=True, check_ex
 
 def get_class_from_type(prov_type):
     """ Serves to translate both provider types and categories to actual classes """
-    all_classes_map = BaseProvider.base_types.copy()
-    for base_type in BaseProvider.base_types.itervalues():
-        all_classes_map.update(base_type.provider_types)
+    all_classes_map = base_types().copy()
```
Comment (RonnyPfannschmidt) [inline utils/providers.py]: this loop and the lookup may need a future rework
please make a card

File: cfme/tests/configure/test_access_control.py [hunk 3]
```diff
@@ -511,16 +511,16 @@ def _go_to(cls, dest='All'):
       {
           'my services': _go_to(MyService),
           'chargeback': _go_to(Server, 'Chargeback'),
-          'clouds providers': _go_to(BaseProvider.base_types['cloud']),
-          'infrastructure providers': _go_to(BaseProvider.base_types['infra']),
+          'clouds providers': _go_to(base_types()['cloud']),
```
Comment (psav) [inline cfme/tests/configure/test_access_control.py]: I think abstracting the entrypoint mechanisms is the right thing to do here.

File: cfme/common/provider.py (hunk 1 above)
Comment (psav) [inline cfme/common/provider.py]: I was going to remove them completely - no one should really be using them anyway, the pass was just to get things working I should actually have removed them.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by psav** [comment]
This is just waiting on the pinning work to go in before it as it will require setup.py installing before usage

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: utils/providers.py [hunk 4]
```diff
@@ -32,6 +32,22 @@
 global_filters = {}
 
 
+def load_setuptools_entrypoints():
+    """ Load modules from querying the specified setuptools entrypoint name.
+        Return the number of loaded plugins. """
```
Comment (mfalesni) [inline utils/providers.py]: This is a lie :) (returns None currently)

File: utils/providers.py [hunk 5]
```diff
@@ -385,9 +401,9 @@ def setup_a_provider_by_class(prov_class, validate=True, check_existing=True, ap
 
 def get_class_from_type(prov_type):
     """ Serves to translate both provider types and categories to actual classes """
-    all_classes_map = BaseProvider.base_types.copy()
-    for base_type in BaseProvider.base_types.itervalues():
-        all_classes_map.update(base_type.provider_types)
+    all_classes_map = base_types().copy()
+    for base_type in base_types().iterkeys():
```
Comment (mfalesni) [inline utils/providers.py]: `.keys` or `six.iterkeys`

**Pull request merged**